pkg/appconsts/appconsts.go

```diff
@@ -115,4 +116,29 @@ var (
 	// that the next message starts at an index that conforms to non-interactive
 	// defaults.
 	NameSpacedPaddedShareBytes = bytes.Repeat([]byte{0}, SparseShareContentSize)
+
+	// FirstCompactShareDataLengthBytes is the number of bytes reserved for the total
+	// data length that is stored in the first compact share of a reserved
+	// namespace. This value is the maximum number of bytes required to store
+	// the data length of a block that only contains compact shares of one type.
+	// For example, if a block contains only evidence then it could contain:
+	// MaxSquareSize * MaxSquareSize * ShareSize bytes of evidence.
+	//
+	// Assuming MaxSquareSize is 128 and ShareSize is 256, this is 4194304 bytes
+	// of evidence. It takes 4 bytes to store a varint of 4194304.
+	//
+	// https://go.dev/play/p/MynwcDHQ_me
+	FirstCompactShareDataLengthBytes = NumberOfBytesVarint(MaxSquareSize * MaxSquareSize * ShareSize)
+
+	// FirstCompactShareContentSize is the number of bytes usable for data in
+	// the first compact share of a reserved namespace. This type of share
+	// contains less space for data than a ContinuationCompactShare because the
+	// first compact share includes a total data length varint.
+	FirstCompactShareContentSize = ContinuationCompactShareContentSize - FirstCompactShareDataLengthBytes
 )
+
+// NumberOfBytesVarint calculates the number of bytes needed to write a varint of n
+func NumberOfBytesVarint(n uint64) (numberOfBytes int) {
+	buf := make([]byte, binary.MaxVarintLen64)
+	return binary.PutUvarint(buf, n)
+}
```

Comment:
it makes sense why we need this here, but I'd prefer to somehow get around having two distinct copies of the same logic (unless we have plans to change the below that I'm unaware of). Could we delete the other code and only use this one, or if we don't want to frequently import a function from a consts package, then we could wrap this function there. wdyt?

https://github.com/celestiaorg/celestia-app/blob/595fd3bc1bdbf9f0ab01a8ef6ce2a2156a7e0cec/pkg/shares/utils.go#L11-L15

Reply:
I opted to duplicate because there is precedence in this repo for duplicating functions when used in distinct package 

Ref:
- https://github.com/rootulp/celestia-app/blob/47a88feadb079c03334f78c6b3e3cf554e4fc1d1/pkg/shares/non_interactive_defaults.go#L80
- https://github.com/rootulp/celestia-app/blob/47a88feadb079c03334f78c6b3e3cf554e4fc1d1/app/estimate_square_size.go#L230

But I'm open to consolidating on one function definition. Using the function defined in the `shares` package introduces an import cycle:

<img width="518" alt="Screen Shot 2022-09-27 at 8 12 12 PM" src="https://user-images.githubusercontent.com/3699047/192659154-412dd8cd-f3c6-42a4-bd13-33f07a960153.png">

so will fallback to the other option proposed: delete the definition in the shares package and use the one in `appconsts`. IMO this isn't ideal because `appconsts` will export a util when I would only expect constants to be defined in it

Reply:
Resolved in https://github.com/celestiaorg/celestia-app/pull/804/commits/285d1a221119d1bca4b7525992bed70e21fb4f37

Reply:
> Using the function defined in the shares package introduces an import cycle:

yeah understood that much :sweat_smile:, I meant to wrap the import in the other direction. or just not export the code in the consts package if we don't want people using it, not unlike the functions linked in the above comment.

Reply:
fwiw, this was likely subconsciously driven by this exact function causing me a full day of debugging raw bytes and hashes

https://github.com/celestiaorg/celestia-app/pull/747/commits/0749b62ecc33a6920c050b90a34bc1ab1f60b60d#r975882452

Reply:
TBH I don't understand the implementation of `DelimLen` that uses `%8` but we have another instance of it here: https://github.com/celestiaorg/celestia-app/blob/595fd3bc1bdbf9f0ab01a8ef6ce2a2156a7e0cec/x/payment/types/payfordata.go#L237-L239

> sorry I meant to wrap the import in the other direction

Sorry I don't understand this approach. Is it what was done in https://github.com/celestiaorg/celestia-app/commit/285d1a221119d1bca4b7525992bed70e21fb4f37 or something else?

> or just not export the code
I'm onboard with this option. Concretely:
1. Revert https://github.com/celestiaorg/celestia-app/commit/285d1a221119d1bca4b7525992bed70e21fb4f37
1. Rename `NumberOfBytesVarint` to `numberOfBytesVarint` in appconsts

Reply:
Resolved in https://github.com/celestiaorg/celestia-app/pull/804/commits/6dc3066a5d7d7816c238d3b7dc6b6c941e0a590b and https://github.com/celestiaorg/celestia-app/pull/804/commits/038e423cccb0c8f38eeb88bdf3cbf2506c818e5d

Reply:
> TBH I don't understand the implementation of DelimLen that uses %8 but we have another instance of it here:

yeah, I don't get it either. thank for finding this!!!! BURN IT WITH FIRE #809